web/src/stream_data.ts

```diff
@@ -259,29 +276,29 @@ export function slug_to_name(slug: string): string {
 
     If there is any ambiguity about whether a stream slug
     is old or modern, we prefer modern, as long as the integer
-    prefix matches a real stream id.  Eventually we will
-    stop supporting the legacy slugs, which only matter now
-    because people have linked to Zulip threads in things like
-    GitHub conversations.  We migrated to modern slugs in
-    early 2018.
+    prefix matches a real stream id. We return undefined if the
+    operand has an unexpected shape, or has the old shape (stream
+    name but no ID) and we don't know of a stream by the given name.
     */
-    const m = /^(\d+)(-.*)?$/.exec(slug);
-    if (m) {
-        const stream_id = Number.parseInt(m[1]!, 10);
-        const sub = sub_store.get(stream_id);
-        if (sub) {
-            return sub.name;
-        }
-        // if nothing was found above, we try to match on the stream
-        // name in the somewhat unlikely event they had a historical
-        // link to a stream like 4-horsemen
+
+    // "New" (2018) format: ${stream_id}-${stream_name} .
+    const match = /^(\d+)(-.*)?$/.exec(slug);
+    const newFormatStreamId = match ? Number.parseInt(match[1]!, 10) : null;
+    if (newFormatStreamId && stream_info.get(newFormatStreamId)) {
+        return newFormatStreamId;
     }
 
-    /*
-    We are dealing with a pre-2018 slug that doesn't have the
-    stream id as a prefix.
-    */
-    return slug;
+    // Old format: just stream name.  This case is relevant indefinitely,
+    // so that links in old conversations (including off-platform like GitHub)
+    // continue to work.
+    const stream = get_sub_by_name(slug);
+    if (stream) {
+        return stream.stream_id;
+    }
+
+    // Unexpected shape, or the old shape and we don't know of a stream with
+    // the given name.
+    return undefined;
```

Comment:
Ah okay, so we do want to return ids even if they might be invalid. Okay I think that's pretty easy to do here!

Reply:
I'm actually unsure if we want do this -- see my comment below in the test file.

Reply:
Navigating to `http://localhost:9991/#narrow/stream/999` either way throws an error, so I just need to separately handle this case. But I think it would be nicer to show the full slug when saying we don't know what the channel is.

Reply:
If the link is in the post-2018 form (which it is by far most likely to be these days), then the slug as a whole is a bit of a computery made-up format, not a user-chosen name — like "2-general" rather than "general". So I'm not sure it's helpful to foreground the slug string anyway.

In particular if we mention the slug as if it's the channel name, that's probably wrong (because the link was probably generated since 2018, in the post-2018 form). If the user followed a link with that operand "2-general" and they lack access, it's probably that there's a channel with ID 2, not a channel with name "2-general". So saying something like "you don't have access to channel '2-general'" is going to be misleading. The best option is probably like "that channel doesn't exist or you lack access", without attempting to say the channel's name.

Reply:
Okay cool, so I'll stick to just saying plainly "Unknown channel" then, since that's simple and less likely to be misleading or incorrect.